File: cli/bun/src/core/service/journal.ts
// The local run journal. Same layout as the Rust product:
// $XDG_STATE_HOME/openprose/cli/production/runs/<session>.json (a dev build
// with a custom endpoint uses cli/custom-<origin digest>/ instead), directories
// 0700, files 0600, entries {session, runId, createdAt, lastSequence,
// sourceSha256} and never a credential; entries older than 30 days are pruned.
import { chmodSync, mkdirSync, openSync, readFileSync, readdirSync, renameSync, rmSync, writeSync, closeSync, fsyncSync } from "node:fs";
import { isAbsolute, join } from "node:path";
import { failure } from "../errors";
import { journalComponent, type Environment, type Json, type JsonObject } from "./manifest";
import { canonicalJson } from "./render";

export interface JournalEntry {
  session: string;
  runId: string | null;
  createdAt: string;
  lastSequence: number;
  sourceSha256: string | null;
}

export function validSession(value: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/u.test(value);
}

export function entryFromJson(value: Json): JournalEntry | undefined {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return undefined;
  const object = value as JsonObject;
  if (typeof object.session !== "string" || !validSession(object.session)) return undefined;
  if (!(object.runId === null || typeof object.runId === "string")) return undefined;
  if (typeof object.createdAt !== "string") return undefined;
  if (typeof object.lastSequence !== "number" || !Number.isSafeInteger(object.lastSequence) || object.lastSequence < 0) return undefined;
  if (!(object.sourceSha256 === null || typeof object.sourceSha256 === "string")) return undefined;
  return { session: object.session, runId: object.runId, createdAt: object.createdAt, lastSequence: object.lastSequence, sourceSha256: object.sourceSha256 };
}

/** An entry's bytes: strict UTF-8 (an entry that is not is skipped, as in Rust) and JSON. */
function entryFromBytes(bytes: Uint8Array): JournalEntry | undefined {
  try { return entryFromJson(JSON.parse(new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(bytes)) as Json); }
  catch { return undefined; }
}

/** Milliseconds of an RFC 3339 date-time (`T`, `t` or a space; `Z`, `z` or an offset), or undefined. */
export function rfc3339(value: string): number | undefined {
  const match = /^(\d{4})-(\d{2})-(\d{2})[Tt ](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$/u.exec(value);
  if (match === null) return undefined;
  const [year, month, day, hour, minute, second] = match.slice(1, 7).map(Number) as [number, number, number, number, number, number];
  const days = [31, year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  // A leap second (:60) counts as the next second, as chrono reads it.
  if (month < 1 || month > 12 || day < 1 || day > days[month - 1]! || hour > 23 || minute > 59 || second > 60) return undefined;
  const zone = match[8]!;
  if (zone.length > 1 && (Number(zone.slice(1, 3)) > 23 || Number(zone.slice(4, 6)) > 59)) return undefined;
  const offset = zone.toUpperCase() === "Z" ? 0 : (zone.startsWith("-") ? -1 : 1) * (Number(zone.slice(1, 3)) * 60 + Number(zone.slice(4, 6)));
  const fraction = match[7] === undefined ? 0 : Number(`0${match[7]}`) * 1000;
  return Date.UTC(year, month - 1, day, hour, minute, second) + fraction - offset * 60_000;
}

function journalFailure(reason: string) {
  return failure("CONFIG_INVALID", { reason: `run journal: ${reason}` });
}

export class Journal {
  constructor(readonly root: string | undefined, readonly directory: string | undefined) {}

  static forEnvironment(env: Readonly<Record<string, string | undefined>>, homeDir: string | undefined, platform: NodeJS.Platform, environment: Environment): Journal {
    const xdg = env.XDG_STATE_HOME;
    let state: string | undefined;
    const localAppData = env.LOCALAPPDATA;
    if (xdg !== undefined && isAbsolute(xdg)) state = xdg;
    // Windows: LOCALAPPDATA first, whether or not a home directory is known.
    else if (platform === "win32" && localAppData !== undefined && localAppData !== "") state = localAppData;
    else if (homeDir !== undefined) {
      state = platform === "darwin" ? join(homeDir, "Library", "Application Support")
        : platform === "win32" ? join(homeDir, "AppData", "Local")
        : join(homeDir, ".local", "state");
    }
    const root = state === undefined ? undefined : join(state, "openprose");
    return new Journal(root, root === undefined ? undefined : join(root, "cli", journalComponent(environment), "runs"));
  }

  private ensure(): string {
    if (this.root === undefined || this.directory === undefined) throw journalFailure("no state directory (set HOME or XDG_STATE_HOME)");
    try { mkdirSync(this.directory, { recursive: true }); } catch { throw journalFailure("cannot create the journal directory"); }
    let current = this.directory;
    for (;;) {
      try { if (process.platform !== "win32") chmodSync(current, 0o700); } catch { throw journalFailure("cannot make the journal directory private"); }
      if (current === this.root) break;
      const parent = join(current, "..");
      if (parent === current) break;
      current = parent;
    }
    return this.directory;
  }

  private path(session: string): string {
    if (!validSession(session)) throw journalFailure("invalid session id");
    if (this.directory === undefined) throw journalFailure("no state directory (set HOME or XDG_STATE_HOME)");
    return join(this.directory, `${session}.json`);
  }

  /** Writes (creates or replaces) an entry atomically with mode 0600. */
  write(entry: JournalEntry): void {
    const directory = this.ensure();
    const path = this.path(entry.session);
    const temporary = join(directory, `.${entry.session}.tmp`);
    // Sorted keys: byte-identical to the Rust product.
    const text = `${canonicalJson({ session: entry.session, runId: entry.runId, createdAt: entry.createdAt, lastSequence: entry.lastSequence, sourceSha256: entry.sourceSha256 })}\n`;
    rmSync(temporary, { force: true });
    let descriptor: number | undefined;
    try {
      descriptor = openSync(temporary, "wx", 0o600);
      writeSync(descriptor, text);
      fsyncSync(descriptor);
    } catch { throw journalFailure("cannot write a journal entry"); }
    finally { if (descriptor !== undefined) closeSync(descriptor); }
    try { renameSync(temporary, path); } catch { throw journalFailure("cannot replace a journal entry"); }
  }

  writeJson(value: Json): void {
    const entry = entryFromJson(value);
    if (entry === undefined) throw journalFailure("invalid preset entry");
    this.write(entry);
  }

  get(session: string): JournalEntry | undefined {
    const path = this.path(session);
    let bytes: Uint8Array;
    try { bytes = readFileSync(path); }
    catch (caught) {
      if ((caught as NodeJS.ErrnoException).code === "ENOENT") return undefined;
      throw journalFailure("cannot read a journal entry");
    }
    return entryFromBytes(bytes);
  }

  /** Every readable entry, newest createdAt first (ties by session). */
  entries(): JournalEntry[] {
    if (this.directory === undefined) return [];
    let names: string[];
    try { names = readdirSync(this.directory); } catch { return []; }
    const entries: JournalEntry[] = [];
    for (const name of names.filter((item) => item.endsWith(".json"))) {
      try {
        const entry = entryFromBytes(readFileSync(join(this.directory, name)));
        if (entry !== undefined) entries.push(entry);
      } catch { /* unreadable entries are ignored */ }
    }
    return entries.sort((left, right) => (right.createdAt < left.createdAt ? -1 : right.createdAt > left.createdAt ? 1 : left.session < right.session ? -1 : left.session > right.session ? 1 : 0));
  }

  findRun(runId: string): JournalEntry | undefined {
    return this.entries().find((entry) => entry.runId === runId);
  }

  /** Removes entries created more than `days` before `now`; only RFC 3339 times count (as in Rust). */
  prune(now: string, days: number): number {
    const reference = rfc3339(now);
    if (reference === undefined) return 0;
    const cutoff = reference - days * 86_400_000;
    let removed = 0;
    for (const entry of this.entries()) {
      const created = rfc3339(entry.createdAt);
      if (created !== undefined && created < cutoff) {
        try { rmSync(this.path(entry.session)); removed += 1; } catch { /* keep */ }
      }
    }
    return removed;
  }
}

File: cli/rust/crates/prose-runner-core/src/service/journal.rs
//! The local run journal.
//!
//! `$XDG_STATE_HOME/openprose/cli/<environment>/runs/<session>.json` (Linux and
//! any platform with `XDG_STATE_HOME`; otherwise `~/.local/state` on Linux,
//! `~/Library/Application Support` on macOS and `%LOCALAPPDATA%` on Windows).
//! `<environment>` is `production` (a `dev-endpoint` build's custom origin:
//! `custom-<origin digest>`). Directories are
//! 0700 and files 0600; entries hold `{session, runId, createdAt, lastSequence,
//! sourceSha256}` and never a credential. Entries older than 30 days are pruned.
use super::Environment;
use crate::{RunnerError, SystemContext};
use serde_json::{Value, json};
use std::path::{Path, PathBuf};

/// One journal entry.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Entry {
    pub session: String,
    pub run_id: Option<String>,
    pub created_at: String,
    pub last_sequence: u64,
    pub source_sha256: Option<String>,
}

impl Entry {
    pub fn to_value(&self) -> Value {
        json!({
            "session": self.session,
            "runId": self.run_id,
            "createdAt": self.created_at,
            "lastSequence": self.last_sequence,
            "sourceSha256": self.source_sha256,
        })
    }

    pub fn from_value(value: &Value) -> Option<Self> {
        let session = value["session"]
            .as_str()
            .filter(|session| valid_session(session))?
            .to_owned();
        Some(Self {
            session,
            run_id: match &value["runId"] {
                Value::Null => None,
                Value::String(run) => Some(run.clone()),
                _ => return None,
            },
            created_at: value["createdAt"].as_str()?.to_owned(),
            last_sequence: value["lastSequence"]
                .as_u64()
                .filter(|sequence| *sequence <= super::http::MAX_SAFE_INTEGER)?,
            source_sha256: match &value["sourceSha256"] {
                Value::Null => None,
                Value::String(digest) => Some(digest.clone()),
                _ => return None,
            },
        })
    }
}

/// A lower-case UUID (the journal file name).
pub fn valid_session(value: &str) -> bool {
    uuid::Uuid::parse_str(value).is_ok_and(|parsed| parsed.hyphenated().to_string() == value)
}

fn failure(reason: &str) -> RunnerError {
    RunnerError::config(format!("run journal: {reason}"))
}

/// The journal directory for one environment.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Journal {
    root: Option<PathBuf>,
    directory: Option<PathBuf>,
}

impl Journal {
    pub fn for_environment(system: &SystemContext, environment: &Environment) -> Self {
        let state = system
            .environment
            .get("XDG_STATE_HOME")
            .filter(|value| Path::new(value).is_absolute())
            .map(PathBuf::from)
            // Windows: LOCALAPPDATA first, whether or not a home directory
            // is known (as in the Bun port).
            .or_else(|| {
                cfg!(windows)
                    .then(|| system.environment.get("LOCALAPPDATA"))
                    .flatten()
                    .filter(|value| !value.is_empty())
                    .map(PathBuf::from)
            })
            .or_else(|| {
                let home = system.home_dir.clone()?;
                Some(if cfg!(target_os = "macos") {
                    home.join("Library").join("Application Support")
                } else if cfg!(windows) {
                    home.join("AppData").join("Local")
                } else {
                    home.join(".local").join("state")
                })
            });
        let root = state.map(|state| state.join("openprose"));
        let directory = root.as_ref().map(|root| {
            root.join("cli")
                .join(environment.journal_component())
                .join("runs")
        });
        Self { root, directory }
    }

    /// A journal rooted at an explicit directory (tests).
    pub fn at(directory: PathBuf) -> Self {
        Self {
            root: Some(directory.clone()),
            directory: Some(directory),
        }
    }

    pub fn directory(&self) -> Option<&Path> {
        self.directory.as_deref()
    }

    fn ensure(&self) -> Result<&Path, RunnerError> {
        let (Some(root), Some(directory)) = (&self.root, &self.directory) else {
            return Err(failure("no state directory (set HOME or XDG_STATE_HOME)"));
        };
        std::fs::create_dir_all(directory)
            .map_err(|_| failure("cannot create the journal directory"))?;
        let mut current = directory.as_path();
        loop {
            private_directory(current)?;
            if current == root.as_path() {
                break;
            }
            match current.parent() {
                Some(parent) => current = parent,
                None => break,
            }
        }
        Ok(directory)
    }

    fn path(&self, session: &str) -> Result<PathBuf, RunnerError> {
        if !valid_session(session) {
            return Err(failure("invalid session id"));
        }
        Ok(self
            .directory
            .as_ref()
            .ok_or_else(|| failure("no state directory (set HOME or XDG_STATE_HOME)"))?
            .join(format!("{session}.json")))
    }

    /// Writes (creates or replaces) an entry atomically with mode 0600.
    pub fn write(&self, entry: &Entry) -> Result<(), RunnerError> {
        let directory = self.ensure()?;
        let path = self.path(&entry.session)?;
        let temporary = directory.join(format!(".{}.tmp", entry.session));
        let mut text = entry.to_value().to_string();
        text.push('\n');
        let _ = std::fs::remove_file(&temporary);
        write_private(&temporary, text.as_bytes())?;
        std::fs::rename(&temporary, &path).map_err(|_| failure("cannot replace a journal entry"))
    }

    /// Writes a raw fixture entry (validated) — test seams only use this.
    pub fn write_value(&self, value: &Value) -> Result<(), RunnerError> {
        let entry = Entry::from_value(value).ok_or_else(|| failure("invalid preset entry"))?;
        self.write(&entry)
    }

    /// Reads one entry by session.
    pub fn get(&self, session: &str) -> Result<Option<Entry>, RunnerError> {
        let path = self.path(session)?;
        match std::fs::read(&path) {
            Ok(bytes) => Ok(serde_json::from_slice::<Value>(&bytes)
                .ok()
                .as_ref()
                .and_then(Entry::from_value)),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(None),
            Err(_) => Err(failure("cannot read a journal entry")),
        }
    }

    /// Every readable entry, newest `createdAt` first (ties by session).
    pub fn entries(&self) -> Vec<Entry> {
        let Some(directory) = &self.directory else {
            return Vec::new();
        };
        let Ok(listing) = std::fs::read_dir(directory) else {
            return Vec::new();
        };
        let mut entries = listing
            .filter_map(Result::ok)
            .filter(|item| item.file_name().to_string_lossy().ends_with(".json"))
            .filter_map(|item| std::fs::read(item.path()).ok())
            .filter_map(|bytes| serde_json::from_slice::<Value>(&bytes).ok())
            .filter_map(|value| Entry::from_value(&value))
            .collect::<Vec<_>>();
        entries.sort_by(|left, right| {
            right
                .created_at
                .cmp(&left.created_at)
                .then_with(|| left.session.cmp(&right.session))
        });
        entries
    }

    /// The newest entry for `run_id`.
    pub fn find_run(&self, run_id: &str) -> Option<Entry> {
        self.entries()
            .into_iter()
            .find(|entry| entry.run_id.as_deref() == Some(run_id))
    }

    /// Removes entries created more than `days` before `now` (RFC 3339).
    pub fn prune(&self, now: &str, days: i64) -> usize {
        let Ok(now) = chrono::DateTime::parse_from_rfc3339(now) else {
            return 0;
        };
        let cutoff = now - chrono::Duration::days(days);
        let mut removed = 0;
        for entry in self.entries() {
            let old = chrono::DateTime::parse_from_rfc3339(&entry.created_at)
                .is_ok_and(|created| created < cutoff);
            if old
                && self
                    .path(&entry.session)
                    .and_then(|path| std::fs::remove_file(path).map_err(|_| failure("remove")))
                    .is_ok()
            {
                removed += 1;
            }
        }
        removed
    }
}

#[cfg(unix)]
fn private_directory(path: &Path) -> Result<(), RunnerError> {
    use std::os::unix::fs::PermissionsExt;
    std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o700))
        .map_err(|_| failure("cannot make the journal directory private"))
}

#[cfg(not(unix))]
fn private_directory(_: &Path) -> Result<(), RunnerError> {
    Ok(())
}

#[cfg(unix)]
fn write_private(path: &Path, bytes: &[u8]) -> Result<(), RunnerError> {
    use std::io::Write;
    use std::os::unix::fs::OpenOptionsExt;
    let mut file = std::fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .mode(0o600)
        .open(path)
        .map_err(|_| failure("cannot create a journal entry"))?;
    file.write_all(bytes)
        .and_then(|()| file.sync_all())
        .map_err(|_| failure("cannot write a journal entry"))
}

#[cfg(not(unix))]
fn write_private(path: &Path, bytes: &[u8]) -> Result<(), RunnerError> {
    use std::io::Write;
    let mut file = std::fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(path)
        .map_err(|_| failure("cannot create a journal entry"))?;
    file.write_all(bytes)
        .map_err(|_| failure("cannot write a journal entry"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn journal_times_match_the_shared_vectors() {
        let fixture: Value = serde_json::from_str(include_str!(
            "../../../../../shared/fixtures/journal-times.json"
        ))
        .unwrap();
        for case in fixture["cases"].as_array().unwrap() {
            let parsed = chrono::DateTime::parse_from_rfc3339(case["value"].as_str().unwrap())
                .ok()
                .map(|time| time.timestamp_millis());
            assert_eq!(serde_json::json!(parsed), case["millis"], "{}", case["id"]);
        }
    }

    fn entry(session: &str, run: Option<&str>, created: &str) -> Entry {
        Entry {
            session: session.into(),
            run_id: run.map(str::to_owned),
            created_at: created.into(),
            last_sequence: 0,
            source_sha256: None,
        }
    }

    #[test]
    fn entries_are_private_and_findable() {
        let directory = tempfile::tempdir().unwrap();
        let journal = Journal::at(directory.path().join("runs"));
        let first = entry(
            "00000000-0000-4000-8000-000000000001",
            Some("run_a"),
            "2026-09-01T00:00:00.000Z",
        );
        let second = entry(
            "00000000-0000-4000-8000-000000000002",
            Some("run_a"),
            "2026-09-02T00:00:00.000Z",
        );
        journal.write(&first).unwrap();
        journal.write(&second).unwrap();
        assert_eq!(journal.find_run("run_a").unwrap().session, second.session);
        assert_eq!(journal.get(&first.session).unwrap().unwrap(), first);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let path = journal
                .directory()
                .unwrap()
                .join(format!("{}.json", first.session));
            assert_eq!(
                std::fs::metadata(path).unwrap().permissions().mode() & 0o777,
                0o600
            );
            assert_eq!(
                std::fs::metadata(journal.directory().unwrap())
                    .unwrap()
                    .permissions()
                    .mode()
                    & 0o777,
                0o700
            );
        }
        assert_eq!(journal.prune("2026-10-01T12:00:00.000Z", 30), 1);
        assert!(journal.get(&first.session).unwrap().is_none());
        assert!(journal.write(&entry("not-a-uuid", None, "x")).is_err());
    }
}
